src/cfi.rs

```diff
@@ -454,12 +456,734 @@ impl<'input, Endian> FrameDescriptionEntry<'input, Endian>
+///         +-------+
+///         | Start |
+///         +-------+
+///             |
+///             |
+/// UninitializedUnwindContext::new()
+///             |
+///             |
+///             V
+/// +----------------------------+
+/// | UninitializedUnwindContext |<---------------.
+/// +----------------------------+                |
+///             |                                 |
+///             |                                 |
+///    ctx.initialize(&cie)              Use with UnwindTable,
+///             |                        and then do ctx.reset()
+///             |                                 |
+///             V                                 |
+///  +--------------------------+                 |
+///  | InitializedUnwindContext |-----------------'
+///  +--------------------------+
+///             |
+///             |
+///            Drop
+///             |
+///             |
+///             V
+///          +-----+
+///          | End |
+///          +-----+
+/// ```
+#[derive(Clone, Debug)]
+pub struct UninitializedUnwindContext<'input, Endian>(UnwindContext<'input, Endian>)
+    where Endian: Endianity;
+
+impl<'input, Endian> UninitializedUnwindContext<'input, Endian>
+    where Endian: Endianity
+{
+    /// Construct a new call frame unwinding context.
+    pub fn new() -> UninitializedUnwindContext<'input, Endian> {
+        UninitializedUnwindContext(UnwindContext::new())
+    }
+
+    /// Run the CIE's initial instructions, creating an
+    /// `InitializedUnwindContext`.
+    pub fn initialize(mut self,
+                      cie: &CommonInformationEntry<'input, Endian>)
+                      -> Result<InitializedUnwindContext<'input, Endian>> {
+        assert!(self.0.is_uninitialized());
+
+        {
+            let mut table = UnwindTable::new_internal(&mut self.0, cie, None);
+            while let Some(_) = try!(table.next_row()) {}
+        }
+
+        self.0.save_initial_rules();
+        Ok(InitializedUnwindContext(self.0))
+    }
+}
+
+/// An initialized unwinding context.
+///
+/// See the documentation for
+/// [`UninitializedUnwindContext`](./struct.UninitializedUnwindContext.html) for
+/// more details.
+#[derive(Clone, Debug)]
+pub struct InitializedUnwindContext<'input, Endian>(UnwindContext<'input, Endian>)
+    where Endian: Endianity;
+
+impl<'input, Endian> InitializedUnwindContext<'input, Endian>
+    where Endian: Endianity
+{
+    /// Reset this context to the uninitialized state.
+    pub fn reset(mut self) -> UninitializedUnwindContext<'input, Endian> {
+        self.0.reset();
+        UninitializedUnwindContext(self.0)
+    }
+}
+
+#[derive(Clone, Debug)]
+enum InitialRules<'input, Endian>
+    where Endian: Endianity
+{
+    // Don't access data in this vec! We just keep it around to reuse the
+    // allocation.
+    Uninitialized(Vec<RegisterRule<'input, Endian>>),
+    // This is safe to access, though.
+    Initialized(Vec<RegisterRule<'input, Endian>>),
+}
```

Comment:
I'm not sure that this enum is carrying its weight. It's not necessary to make the module API safe to use (`InitializedUnwindContext` and `UninitializedUnwindContext` are enough for that), and it's creating a lot of complexity regarding `UnwindContext::initial_rules`.